### PR TITLE
Ability to interrupt/suspend reading without having to disable auto r…

### PR DESCRIPTION
**Pull request by rkapsi**
…eading

ATTENTION: This is a preliminary PR to see if it's feasible and gather some early feedback. I'll provide Unit Tests and say exploits described in #8953 in a subsequent PR if you like this idea.

Motivation

There are to aspects to this change. First: There is no way to break Netty out of an ongoing read() operation other than disabling auto reading. Second: The auto read flag is a "per connection global" mutable state. Working with it in a complex Netty application with many handlers can be awkward because anyone could re-enable it at any time and wheather a hander that disabled it is ready or not.

Modifications

Introduce an interruptReading() method which allows the user to break out of an ongoing read() operation and suspend further reading until Channel#read() is called again. Each handler in the pipeline has the opportunity to intercept and suppress attempts to produce more data if they're not ready it.

Result

Implements #8953

### CODE REVIEW

**Comment by normanmaurer** [comment]
@ejona86 @trustin @carl-mastrangelo @mosesn @bryce-anderson @njhill PTAL as well.

**Comment by rkapsi** [comment]
@normanmaurer I have added an unit test that highlights the problematic and how this helps. I will squash the commits further down the road.

Some open questions:

Do you like the method name? It's `interruptReading()` right now but it could be `interruptRead()`, `suspendX()`, `yieldX()`?

Is `ChannelConfig` the right place for it? I added it there just to get something working but `Channel` seems a more appropriate place for it.

The interrupt flag is currently backed by an `AtomicReferenceFieldUpdater`. I did it just to maintain symmetry to the autoRead flag but I believe it can be just a `volatile boolean`.

**Comment by ejona86** [comment]
I don't quite understand how having something like this on config makes things much better. But I've been pushing to move writability to the channelhandler/ctx for Netty 5, and interruptReading seems like it would fit in well too. Although that makes it a Netty 5 "thing." Flow control really needs to be handler-to-handler.

**Comment by rkapsi** [comment]
@ejona86 ack regards config. I'm looking for suggestions. My thought is `Channel` or maybe even bury it as a public method in`AbstractChannel` for 4.1 and ask the user to cast until a more appropriate/permanent but potentially API breaking location can be found (i.e. 5.0)?

Ack also regards handler-to-handler flow control. This PR is just focused on this one aspect. We solve the flow-control problem with a custom handler which maintains an internal queue/buffer and hands out messages when asked to do so and reads more messages if the queue gets empty. 

It's an evolution of the [FlowControlHandler](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/flow/FlowControlHandler.java) which I contributed a few years back and I can maybe revamp it as a follow up and switch it from auto reading to this (it'd be slightly api breaking though). We stopped using FlowControlHandler ourselves because it doesn't work well with multiple instances of it in the pipeline (they screw up the auto reading).

**Comment by rkapsi** [comment]
@ejona86 @normanmaurer please let me know what you think of the most recent push? I'm introducing an `Interruptible` interface which all Channel types implement that were previously dependent on `DefaultChannelConfig#autoReadCleared()`. This removes the method I originally added and gets rid of a handful of methods such as `KQueueChannelConfig#autoReadCleared()`.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollChannel.java [hunk 1]
```diff
@@ -244,6 +247,22 @@ private static boolean isAllowHalfClosure(ChannelConfig config) {
...
+        return true;
+      }
+      return false;
+    }
+
+    @Override
+    public final void interrupt() {
+        if (!interrupted) {
+          interrupted = true;
+          clearEpollIn();
```
Comment (rkapsi) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollChannel.java]: This method and its equivalent buddies in the other Channel implementations are either protected or package private methods as they needed to be accessible from the various ChannelConfig implementations. With this change they can be made private.

**Comment by rkapsi** [comment]
Friendly poke @normanmaurer.

**Comment by normanmaurer** [comment]
@rkapsi sorry I am a bit under the water atm :/ Will try to check tomorrow

**Comment by normanmaurer** [comment]
@rkapsi so after checking the code and think a bit about this I would vote to postpone this to netty 5. There we can fix all of this in a more elegant way. This is also not the same but related to this PR that I closed:

https://github.com/netty/netty/pull/9014

**Comment by trustin** [comment]
Not sure it will be helpful, but this problem can be solved by having a single place that counts the number of pending bytes. For example, something like this:

- https://github.com/line/armeria/blob/armeria-0.83.0/core/src/main/java/com/linecorp/armeria/internal/InboundTrafficController.java (Enables or disables auto-read based on the number of unconsumed inbound bytes.)
- https://github.com/line/armeria/blob/armeria-0.83.0/core/src/main/java/com/linecorp/armeria/internal/ReadSuppressingHandler.java (Suppresses all `ctx.read()` if auto-read is disabled.)

This will not work if someone in the pipeline calls `setAutoRead()` arbitrarily rather than going through `InboundTrafficController`, but I think it's not very likely.

**Comment by trustin** [comment]
Inbound traffic control is trickier than outbound traffic control because Netty cannot determine 'when' the inbound data is consumed, which requires explicit feed back from user code, i.e. Netty can call `InboundTrafficController.inc()` for a user, but a user has to call `InboundTrafficController.dec()` because it is highly application-dependent. That was why I have been reluctant to the ideas related with implementing inbound traffic control inside Netty, but it's be nice if there's a good fix for this.

In this PR, it seems to me that `interrupt()` is not orthogonal with `setAutoRead()`. What would happen if a user calls `interrupt()` when auto-read is on or off, or vice versa? I think it is not very clear for a user to guess what will happen in every case. Can we streamline this further? (Not that I have a good answer. 😅)

**Comment by rkapsi** [comment]
Hey there,

sorry for the silence. I had to fiddle with something Netty related. I'm OK moving this to Netty 5 as I can kind of get away with the existing auto reading flag in 4.1. It's not ideal though and this would be much nicer.

The heart of this idea is to provide a mechanism to break out of [this loop](https://github.com/netty/netty/blob/00afb19d7a37de21b35ce4f6cb3fa7f74809f2ab/transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java#L146-L165). I'm referring to NIO just symbolically. The same loop exists in the other Channel implementations.

With auto reading enabled it will consume bytes from the wire and `fireChannelRead(...)` will fire for as long as auto reading is true. The only way to break out of it is to set auto reading to false. It's fine with one handler that does it but gets tricky if you have many of them. It becomes a coordination problem in respect to who turned it off and/or on.

I'd be much nicer if I can "interrupt" the loop (only to be called from `channelRead(...)`) and then suppress the `ctx.read()`. Each handler can do it in their own little universe managing their own state. If they need more data they call `ctx.read()`. If they need to put things on hold they interrupt out of `channelRead(...)` and suppress `ctx.read()` coming from other handlers higher up the stack.

For proper flow control you'll also need to implement some kind of queueing in your handler as one `read()` of `ByteBuf` can be easily decoded into a few hundred let's say HttpRequests (think HTTP piplinging and exploiting it with malicious intent). 

Does this make sense?

**Comment by carl-mastrangelo** [comment]
I have a dumb question: Why not use plain old Thread.interrupted()?

**Comment by rkapsi** [comment]
I don't think it'd be viable. `Thread#interrupt()` is for interrupting Threads while many Channels share one I/O Thread. One of Netty's core concepts. If anyone calls `Thread#Interrupt()` outside of `channelRead(...)` you'll end up messing with some random Channel. The other problem is that handlers may execute on a different Thread than the Channel's I/O Thread. Lastly, anyone can clear the interrupted flag by mere reading it and it subsequently having no effect.

**Pull request closed without merging**